bots/notification-bot.ts

```diff
@@ -0,0 +1,246 @@
+
+  // If we got here, all retries failed
+  throw new Error(
+    `Failed to send push notifications after ${MAX_RETRIES} attempts. Last error: ${lastError?.message}`,
+  );
+}
+
+/**
+ * Handles a Communication resource change event.
+ * Sends push notifications to relevant recipients based on the message content.
+ * @param medplum - The Medplum client.
+ * @param event - The bot event.
+ */
+export async function handler(medplum: MedplumClient, event: BotEvent): Promise<void> {
+  // Get the Communication resource from the event
+  const communication = event.input as Communication;
+
+  // Skip if this is a thread (no partOf reference)
+  if (!communication.partOf?.[0]) {
+    return;
+  }
+
+  // Skip if this is not a new message
+  if (communication.received || communication.status === "completed") {
+    return;
+  }
+
+  // Get the thread
+  const thread = await medplum.readReference(communication.partOf[0]);
+  if (!thread || thread.resourceType !== "Communication") {
+    return;
+  }
+
+  // Get the patient (subject of the thread)
+  const patient = await medplum.readReference(thread.subject!);
+  if (!patient || patient.resourceType !== "Patient") {
+    return;
+  }
+
+  // Get the sender
+  const sender = await medplum.readReference(communication.sender!);
+  if (!sender || (sender.resourceType !== "Patient" && sender.resourceType !== "Practitioner")) {
+    return;
+  }
+
+  // Determine recipients based on sender type
+  let recipients: (Patient | Practitioner)[] = [];
+  if (sender.resourceType === "Patient") {
+    // If sender is patient, notify practitioners
+    const practitioners = await medplum.searchResources("Practitioner", {});
+    recipients = practitioners;
+  } else if (sender.resourceType === "Practitioner") {
+    // If sender is practitioner, notify the patient
+    recipients = [patient];
+  }
+
+  // Get message content
+  const hasAttachment = communication.payload?.some((p) => p.contentAttachment);
+
+  // Create notifications for all recipients
+  const notifications: ExpoNotificationMessage[] = [];
+  for (const recipient of recipients) {
+    const pushToken = recipient.extension?.find(
+      (e) => e.url === "https://medplum.com/push-token",
+    )?.valueString;
+
+    if (pushToken) {
+      notifications.push({
+        to: pushToken,
+        title: `New message from ${getSenderDisplayName(sender)}`,
+        body: hasAttachment ? "📎 Click to view attachment" : "Click to view message",
+        data: {
+          threadId: thread.id,
+        },
+        sound: "default",
+        badge: 1,
+      });
+    }
+  }
+
+  // Send notifications in batches of 100
+  for (let i = 0; i < notifications.length; i += BATCH_SIZE) {
+    const batch = notifications.slice(i, i + BATCH_SIZE);
+    try {
+      await sendNotificationBatch(batch);
+    } catch (error) {
+      console.error(`Error sending batch ${i / BATCH_SIZE + 1}:`, error);
+      // Continue with next batch even if this one failed
+    }
+  }
```

Comment:
Should we use `Promise.all`? Do we need to wait each batch to be finished to start the next one?

Reply:
Yes due to rate limiting. And I find it hard to send more than 100 notifications at once, so that's probably enough for a good time from a performance PoV.